File: leetcode-cn/40.组合总和-ii.py
#
# @lc app=leetcode.cn id=40 lang=python3
#
# [40] 组合总和 II
#

# @lc code=start
from typing import List


class Solution:
    def combinationSum2(self, candidates: List[int],
                        target: int) -> List[List[int]]:
        path = []
        ans = []

        def help(target, candidates):
            if target == 0:
                ans.append(path[:])

            for i, num in enumerate(candidates):
                if i > 0 and candidates[i] == candidates[i - 1]:
                    continue
                if num > target:
                    break

                path.append(num)
                help(target - num, candidates[i + 1:])
                path.pop()

        candidates.sort()
        help(target, candidates)
        return ans


# @lc code=end

File: leetcode-cn/C++/40.组合总和-ii.cpp
/*
 * @lc app=leetcode.cn id=40 lang=cpp
 *
 * [40] 组合总和 II
 */

// @lc code=start
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> path;
    vector<vector<int>> ans;
    vector<int> candidates;

    void help(int index, int target)
    {
        if (target == 0) {
            ans.push_back(path);
            return;
        }

        for (int i = index; i < candidates.size(); i++) {
            if (i > index && candidates[i] == candidates[i - 1]) {
                continue;
            }

            if (candidates[i] > target) {
                break;
            }

            path.push_back(candidates[i]);
            help(i + 1, target - candidates[i]);
            path.pop_back();
        }
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target)
    {
        sort(candidates.begin(), candidates.end());
        this->candidates = candidates;

        help(0, target);
        return ans;
    }
};
// @lc code=end
